File: backend/app/routers/reports.py
"""
Reports API Router
Analytics and reporting endpoints - Using Raw SQL
"""
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from sqlalchemy import text
from app.database.sql import get_db
from app.database.mongodb import get_activity_logs_collection
from datetime import datetime, timedelta
from typing import List, Dict, Any

router = APIRouter(prefix="/reports", tags=["reports"])


@router.get("/activity-logs")
def get_activity_logs(
    limit: int = 100,
    skip: int = 0
):
    """Get recent activity logs from MongoDB"""
    try:
        logs_collection = get_activity_logs_collection()
        if logs_collection is None:
            return {"error": "MongoDB not connected", "logs": []}
        
        logs = list(logs_collection.find(
            {},
            {"_id": 0}
        ).sort("timestamp", -1).skip(skip).limit(limit))
        
        # Convert datetime to string for JSON serialization
        for log in logs:
            if "timestamp" in log:
                log["timestamp"] = log["timestamp"].isoformat()
        
        return {"logs": logs, "total": logs_collection.count_documents({})}
    except Exception as e:
        return {"error": str(e), "logs": []}


@router.get("/circulation-stats")
def get_circulation_stats(db: Session = Depends(get_db)):
    """Get circulation statistics"""
    now = datetime.utcnow()
    
    # Today's checkouts
    today_start = datetime(now.year, now.month, now.day)
    result = db.execute(
        text("SELECT COUNT(*) as count FROM transactions WHERE checkout_date >= :today_start"),
        {"today_start": today_start}
    )
    today_checkouts = result.fetchone().count
    
    # This week's checkouts
    week_start = now - timedelta(days=now.weekday())
    result = db.execute(
        text("SELECT COUNT(*) as count FROM transactions WHERE checkout_date >= :week_start"),
        {"week_start": week_start}
    )
    week_checkouts = result.fetchone().count
    
    # This month's checkouts
    month_start = datetime(now.year, now.month, 1)
    result = db.execute(
        text("SELECT COUNT(*) as count FROM transactions WHERE checkout_date >= :month_start"),
        {"month_start": month_start}
    )
    month_checkouts = result.fetchone().count
    
    # Average checkout duration
    result = db.execute(
        text("""
            SELECT 
                AVG(EXTRACT(EPOCH FROM (return_date - checkout_date))/86400) as avg_duration
            FROM transactions 
            WHERE return_date IS NOT NULL
        """)
    )
    avg_row = result.fetchone()
    avg_duration = avg_row.avg_duration if avg_row.avg_duration else 0
    
    return {
        "today_checkouts": today_checkouts,
        "week_checkouts": week_checkouts,
        "month_checkouts": month_checkouts,
        "average_checkout_duration_days": round(avg_duration, 2)
    }


@router.get("/popular-books")
def get_popular_books(limit: int = 10, db: Session = Depends(get_db)):
    """Get most borrowed books"""
    result = db.execute(
        text("""
            SELECT 
                b.id,
                b.title,
                b.author,
                COUNT(t.id) as borrow_count
            FROM books b
            JOIN transactions t ON b.id = t.book_id
            GROUP BY b.id, b.title, b.author
            ORDER BY borrow_count DESC
            LIMIT :limit
        """),
        {"limit": limit}
    )
    popular = result.fetchall()
    
    return [
        {
            "book_id": row.id,
            "title": row.title,
            "author": row.author,
            "borrow_count": row.borrow_count
        }
        for row in popular
    ]


@router.get("/category-distribution")
def get_category_distribution(db: Session = Depends(get_db)):
    """Get book distribution by category"""
    result = db.execute(
        text("""
            SELECT 
                category,
                COUNT(id) as count
            FROM books
            GROUP BY category
        """)
    )
    categories = result.fetchall()
    
    return [
        {"category": cat.category or "Uncategorized", "count": cat.count}
        for cat in categories
    ]


@router.get("/member-stats")
def get_member_stats(db: Session = Depends(get_db)):
    """Get member statistics by type"""
    result = db.execute(
        text("""
            SELECT 
                member_type,
                COUNT(id) as count
            FROM members
            GROUP BY member_type
        """)
    )
    member_types = result.fetchall()
    
    result = db.execute(
        text("SELECT COUNT(*) as count FROM members WHERE status = 'active'")
    )
    active_members = result.fetchone().count
    
    return {
        "by_type": [
            {"type": mt.member_type, "count": mt.count}
            for mt in member_types
        ],
        "active_members": active_members
    }


@router.get("/fine-report")
def get_fine_report(db: Session = Depends(get_db)):
    """Get fine collection report"""
    # Total outstanding fines
    result = db.execute(
        text("SELECT SUM(fines) as total FROM members")
    )
    outstanding = result.fetchone().total or 0
    
    # Collected fines (from paid transactions)
    result = db.execute(
        text("SELECT SUM(fine_amount) as total FROM transactions WHERE fine_paid = TRUE")
    )
    collected = result.fetchone().total or 0
    
    # Unpaid fines
    result = db.execute(
        text("SELECT SUM(fine_amount) as total FROM transactions WHERE fine_paid = FALSE AND fine_amount > 0")
    )
    unpaid = result.fetchone().total or 0
    
    # Members with fines
    result = db.execute(
        text("SELECT COUNT(*) as count FROM members WHERE fines > 0")
    )
    members_with_fines = result.fetchone().count
    
    return {
        "outstanding_fines": round(outstanding, 2),
        "collected_fines": round(collected, 2),
        "unpaid_fines": round(unpaid, 2),
        "members_with_fines": members_with_fines
    }


@router.get("/monthly-trend")
def get_monthly_trend(db: Session = Depends(get_db)):
    """Get monthly checkout trend for the last 6 months"""
    now = datetime.utcnow()
    six_months_ago = now - timedelta(days=180)
    
    # PostgreSQL date extraction
    result = db.execute(
        text("""
            SELECT 
                EXTRACT(YEAR FROM checkout_date)::INTEGER as year,
                EXTRACT(MONTH FROM checkout_date)::INTEGER as month,
                COUNT(id) as count
            FROM transactions
            WHERE checkout_date >= :six_months_ago
            GROUP BY year, month
            ORDER BY year, month
        """),
        {"six_months_ago": six_months_ago}
    )
    monthly_data = result.fetchall()
    
    return [
        {
            "year": row.year,
            "month": row.month,
            "checkouts": row.count
        }
        for row in monthly_data
    ]

File: backend/src/routes/reports.js
import express from 'express';
import prisma from '../db/prisma.js';
import { getActivityLogsWithTotal } from '../db/activityLogger.js';
import { withCache } from '../utils/cache.js';
import { parseSkipLimitPagination } from '../utils/pagination.js';

const router = express.Router();
const REPORTS_CACHE_MS = Number.parseInt(process.env.REPORTS_CACHE_MS || '30000', 10);

router.get('/activity-logs', async (req, res) => {
    try {
        const { limit, skip } = parseSkipLimitPagination(req.query, {
            defaultLimit: 50,
            maxLimit: 200,
            maxSkip: 5000,
        });

        const { logs, total } = await getActivityLogsWithTotal(limit, skip);
        const serializedLogs = logs.map((log) => ({
            ...log,
            timestamp: log.timestamp instanceof Date
                ? log.timestamp.toISOString()
                : log.timestamp,
        }));

        res.json({
            logs: serializedLogs,
            total,
            limit,
            skip,
        });
    } catch (error) {
        res.status(500).json({ error: error.message, logs: [] });
    }
});

router.get('/circulation-stats', async (req, res) => {
    try {
        const payload = await withCache('reports:circulation-stats', REPORTS_CACHE_MS, async () => {
            const [totalCheckouts, totalReturns, avgDuration] = await Promise.all([
                prisma.transactions.count(),
                prisma.transactions.count({ where: { status: 'returned' } }),
                prisma.$queryRaw`
                    SELECT AVG(EXTRACT(EPOCH FROM (return_date - checkout_date)) / 86400) as avg_days
                    FROM transactions WHERE return_date IS NOT NULL
                `,
            ]);

            return {
                total_checkouts: totalCheckouts,
                total_returns: totalReturns,
                average_checkout_duration_days: avgDuration[0]?.avg_days
                    ? parseFloat(avgDuration[0].avg_days).toFixed(1)
                    : 0,
            };
        });

        res.json(payload);
    } catch (error) {
        res.status(500).json({ detail: error.message });
    }
});

router.get('/popular-books', async (req, res) => {
    try {
        const { limit } = parseSkipLimitPagination(
            { limit: req.query.limit, skip: 0 },
            { defaultLimit: 10, maxLimit: 50, maxSkip: 0 }
        );

        const payload = await withCache(`reports:popular-books:${limit}`, REPORTS_CACHE_MS, async () => (
            prisma.$queryRaw`
                SELECT b.id, b.title, b.author, b.isbn, COUNT(t.id)::int as checkout_count
                FROM books b
                JOIN transactions t ON b.id = t.book_id
                GROUP BY b.id, b.title, b.author, b.isbn
                ORDER BY checkout_count DESC
                LIMIT ${limit}
            `
        ));

        res.json(payload);
    } catch (error) {
        res.status(500).json({ detail: error.message });
    }
});

router.get('/category-distribution', async (req, res) => {
    try {
        const payload = await withCache('reports:category-distribution', REPORTS_CACHE_MS, async () => {
            const categories = await prisma.books.groupBy({
                by: ['category'],
                _count: { id: true },
                orderBy: { _count: { id: 'desc' } },
            });

            return categories.map((category) => ({
                category: category.category || 'Uncategorized',
                count: category._count.id.toString(),
            }));
        });

        res.json(payload);
    } catch (error) {
        res.status(500).json({ detail: error.message });
    }
});

router.get('/member-stats', async (req, res) => {
    try {
        const payload = await withCache('reports:member-stats', REPORTS_CACHE_MS, async () => {
            const stats = await prisma.members.groupBy({
                by: ['member_type'],
                _count: { id: true },
            });

            const statusStats = await prisma.members.groupBy({
                by: ['status'],
                _count: { id: true },
            });

            return {
                by_type: stats.map((entry) => ({
                    member_type: entry.member_type,
                    count: entry._count.id.toString(),
                })),
                by_status: statusStats.map((entry) => ({
                    status: entry.status,
                    count: entry._count.id.toString(),
                })),
            };
        });

        res.json(payload);
    } catch (error) {
        res.status(500).json({ detail: error.message });
    }
});

const getFineReportPayload = async () => withCache('reports:fines', REPORTS_CACHE_MS, async () => {
    const [totalFines, unpaidFines, membersWithFines] = await Promise.all([
        prisma.members.aggregate({ _sum: { fines: true } }),
        prisma.transactions.aggregate({
            _sum: { fine_amount: true },
            where: { fine_paid: false, fine_amount: { gt: 0 } },
        }),
        prisma.members.findMany({
            where: { fines: { gt: 0 } },
            select: { id: true, name: true, email: true, fines: true },
            orderBy: { fines: 'desc' },
            take: 20,
        }),
    ]);

    return {
        total_fines: totalFines._sum.fines || 0,
        unpaid_fines: unpaidFines._sum.fine_amount || 0,
        members_with_fines: membersWithFines,
    };
});

router.get('/fines', async (req, res) => {
    try {
        const payload = await getFineReportPayload();
        res.json(payload);
    } catch (error) {
        res.status(500).json({ detail: error.message });
    }
});

// Backward-compatible endpoint used by existing frontend service.
router.get('/fine-report', async (req, res) => {
    try {
        const payload = await getFineReportPayload();
        res.json(payload);
    } catch (error) {
        res.status(500).json({ detail: error.message });
    }
});

router.get('/monthly-trend', async (req, res) => {
    try {
        const payload = await withCache('reports:monthly-trend', REPORTS_CACHE_MS, async () => (
            prisma.$queryRaw`
                SELECT
                    EXTRACT(YEAR FROM checkout_date)::int as year,
                    EXTRACT(MONTH FROM checkout_date)::int as month,
                    COUNT(*)::int as checkouts
                FROM transactions
                GROUP BY year, month
                ORDER BY year DESC, month DESC
                LIMIT 12
            `
        ));

        res.json(payload);
    } catch (error) {
        res.status(500).json({ detail: error.message });
    }
});

export default router;
